packages/grid/_modules_/grid/hooks/features/selection/useGridSelection.ts

```diff
@@ -94,8 +98,16 @@ export const useGridSelection = (apiRef: GridApiRef): void => {
         data: row,
         isSelected: selectionState[id] !== undefined,
       };
+
+      let newSelectionModel = Object.values(selectionState);
+      if (selectionModel) {
+        newSelectionModel = selectionModel.includes(rowModelParams.id)
+          ? selectionModel.filter((selection) => selection !== rowModelParams.id)
+          : [...selectionModel, rowModelParams.id];
+      }
```

Comment:
Wait, are we implementing the logic that is inside the above `setGridState` twice?

Reply:
No, it looks like it but it is much simpler. That one only appends the newly selected row id to the already existing selection model so that the value of the params that the developer will get from the `onSelectionModelChange` is correct. There is no state or options update.

Reply:
I don't understand, want what about multi-selection? It seems that it should be exactly the same, so enforced with a shared abstraction

Reply:
We use an object for the selection state to avoid looping through an array to know if a row is selected, otherwise, if you have 1000k rows we would have a perf hit. 
https://github.com/mui-org/material-ui-x/blob/master/packages/grid/_modules_/grid/components/GridViewport.tsx#L80

However, for the prop, using an object is an option but it seemed to make less sense, and it is more convenient to use an array.

Reply:
@dtassone The caching makes sense, it definitely seems valid. What do you think about we stored the controlled prop as is, in the store, and store a private cache, a dedicated state, in another part of the state (one for instance that we wouldn't expose for developers than want to store the state in their data base). But more importantly, have updates on the `state.selectionModal` being the source of truth, and triggering changes automatically, to its cache, say `state.cache.selectionModal`?

Reply:
Well the controlled prop to me is already stored and is the value of the prop itself. Why would we need to duplicate the values? 
And the private cache is actually the grid state and it is the source of truth. 
We could do a selector that transforms the state to make it more storable.

Reply:
> Why would we need to duplicate the values?

To fix the first comment of this thread.

Reply:
Actually, I was looking in that use case yesterday and what I struggle to understand is how would you differentiate multiselection using `CTRL` from the normal one? The only way I see it is that if it is not multi-selection then basically every time the `onSelectionModelChange` is called we need to always pass just one ID (the lasted one)?

Then for the muti-selection and checkbox selection, you would append/remove IDs based on what is being selected/deselected.

Reply:
Considering #1823 and the equality issue and the discussion here, how about we transform state to be optimised in a selector. As they are memoised it will only do the transformation when the underlying model change. That way it would allow us to store the model in state, fix the equality issue #1823 and avoid the model duplication.
The cache mentioned by @oliviertassinari would actually be provided by the selector memoization. What do you think?

Reply:
@dtassone This could work if we also expose the raw state. We need the raw state to perform the control feedback loop.